tensorflow/contrib/avro/ops/parse_avro_record.cc

```diff
@@ -0,0 +1,2058 @@
+         c->set_output(output_idx++, c->Vector(c->UnknownDim()));
+       }
+       for (int i_sparse = 0; i_sparse < attrs.num_sparse; ++i_sparse) {
+         c->set_output(output_idx++, c->Vector(2));
+       }
+
+       // Output dense_values
+       for (int i_dense = 0; i_dense < attrs.num_dense; ++i_dense) {
+         ShapeHandle dense;
+         TF_RETURN_IF_ERROR(c->MakeShapeFromPartialTensorShape(
+             attrs.dense_infos[i_dense].shape, &dense));
+         TF_RETURN_IF_ERROR(c->Concatenate(input, dense, &dense));
+         c->set_output(output_idx++, dense);
+       }
+
+       return Status::OK();
+     })
+    .Doc(R"doc(
+      Parses a serialized avro record that follows the supplied schema into typed tensors.
+      serialized: A vector containing a batch of binary serialized avro records.
+      dense_keys: A list of Ndense string Tensors.
+        The keys expected are associated with dense values.
+      dense_defaults: A list of Ndense Tensors (some may be empty).
+        These defaults can either be fully defined for all values in the tensor or they
+        can be defined as padding element that is used whenever the original input data
+        misses values to fill out the full tensor.
+      dense_shapes: A list of Ndense shapes; the shapes of the dense tensors.
+        The number of elements corresponding to dense_key[j]
+        must always equal dense_shapes[j].NumEntries().
+        If dense_shapes[j] == (D0, D1, ..., DN) then the shape of output
+        Tensor dense_values[j] will be (|serialized|, D0, D1, ..., DN):
+        The dense outputs are just the inputs row-stacked by batch.
+        This works for dense_shapes[j] = (-1, D1, ..., DN).  In this case
+        the shape of the output Tensor dense_values[j] will be
+        (|serialized|, M, D1, .., DN), where M is the maximum number of blocks
+        of elements of length D1 * .... * DN, across all minibatch entries
+        in the input.  Any minibatch entry with less than M blocks of elements of
+        length D1 * ... * DN will be padded with the corresponding default_value
+        scalar element along the second dimension.
+      dense_types: A list of Ndense types; the type of data in each Feature given in dense_keys.
+      sparse_keys: A list of Ndense string Tensors (scalars).
+        The keys expected are associated with sparse values.
+      sparse_types: A list of Nsparse types; the data types of data in each Feature
+        given in sparse_keys.
+      schema: A string that describes the avro schema of the underlying serialized avro string.
+        Currently the parse function supports the primitive types DT_STRING, DT_DOUBLE, DT_FLOAT,
+        DT_INT64, DT_INT32, and DT_BOOL.
+        The supported avro version depends on the compiled library avro library linked against
+        TensorFlow during build. More instructions on avro:
+        https://avro.apache.org/docs/1.8.1/spec.html
+    )doc");
```

Comment:
Put the REGISTER_OP at the top of the file, since it's good documentation for what's implemented in it.

Reply:
Sorry, we can't move this to the top w/out breaking out some of the class definitions/implementations into separate files.